zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -213,19 +214,19 @@ public void afterExecutionPathCompleted(
       }
       return;
     }
+    final var inputCollectionOrFailure = readInputCollectionVariable(element, flowScopeContext);
+    if (inputCollectionOrFailure.isLeft()) {
+      // this incident is un-resolvable
+      incidentBehavior.createIncident(inputCollectionOrFailure.getLeft(), childContext);
+      return;
+    }
 
-    if (loopCharacteristics.isSequential()) {
-
-      final var inputCollectionOrFailure = readInputCollectionVariable(element, flowScopeContext);
```

Comment:
❓  This may come with a performance hit for parallel multi-instance, as it now reads the input collection variable on each inner instance completion. Did we check the performance of this?

💭 Note that reading the variable here means evaluating the input collection expression again. Most cases will probably do a direct lookup of a list variable, but theoretically this expression could require multiple variable lookups.

Reply:
I see your point. No, we didn't check any performance implications, we focused more on resolving the bug.

We indeed `readInputCollectionVariable` at `onActivate` and `onActivating` and we could use these times to get the input collection size, I've looked around now and I could't decide where/how to store it that it keeps it relevant to an element. 

The elementInstance state could be one candidate, but I feel like the `MultiInstanceBodyProcessor` is not the right place to alter the state.

Maybe I could use `stateBehavior.setLocalVariable()`? But I'm also not sure if this is a correct use case to use it.

Any suggestions?

Reply:
It would be new information for an `ElementInstance` that is not yet tracked. Something along the lines of `childPlannedCount` or `multiInstanceInputSize`, which we could add next to the existing `childCount`, `childActivatedCount`, `childCompletedCount`, and `childTerminatedCount`:

https://github.com/camunda/camunda/blob/2d623b82dee3dfb8ee3b8c8fe89d03acccf5696b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/ElementInstance.java#L28-L34

We can store this on `ProcessInstance:ELEMENT_ACTIVATING` or `ProcessInstance:ELEMENT_ACTIVATED` by adding a new property on it. However, this property would be specific to multi-instance body element instances, so all other records always get the same value (1) which is quite wasteful (at least a few extra bytes for each process instance record).

Alternatively, we can introduce a new record/event to store this specific data, e.g. something along the lines of `MultiInstance:BODY_ACTIVATING`. It can reference the multi-instance body by key, and provide some additional information:
- input collection size: important for this
- input collection itself: I don't see a use for this yet, but might be helpful later (I'd not add this if we have no need for it yet)

Perhaps there are other records we can use to store this data. But this is what comes to mind. Curious to your thoughts about this @mustafadagher @remcowesterhoud.

Reply:
@korthout Do you mean as in to add it in the scope of `EventAppliers`? I don't think we have the `expressionBehavior` or any of the `BpmnBehaviors` available for us in this phase to evaluate the expression and store the result.

Reply:
>>Alternatively, we can introduce a new record/event to store this specific data, e.g. something along the lines of MultiInstance:BODY_ACTIVATING. It can reference the multi-instance body by key, and provide some additional information

>Fetching data from the state versus evaluating an expression. I wonder what's more performant 🤔 We do already evaluate an input expression if the MI-body is sequential as opposed to parallel. But I suppose performance may be less of an issue if the MI is sequential anyways.

I meant that the event can be used to fill the data in the `ElementInstance`. It would not require an additional lookup on completion of an inner instance as the `ElementInstance` of the multi-instance body is already looked up.

https://github.com/camunda/camunda/blob/2d623b82dee3dfb8ee3b8c8fe89d03acccf5696b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java#L226-L228

> >We can store this on ProcessInstance:ELEMENT_ACTIVATING or ProcessInstance:ELEMENT_ACTIVATED by adding a new property on it. However, this property would be specific to multi-instance body element instances, so all other records always get the same value (1) which is quite wasteful (at least a few extra bytes for each process instance record).

>I would prefer this. We already store the other properties (childActivatedCount, childCompletedCount) which are also not relevant for many of the elemant instances

These are incremented by the `ELEMENT_ACTIVATED` and `ELEMENT_COMPLETED` appliers of the inner instances. No additional information is stored for them in the `ProcessInstance` records.

> I would argue the default value would be 0 as for most elements no children are planned.

👍 You're right.

Reply:
One other way to do it could be: the output collection has the same size as the input collection, we already initialize it and store it as a `Local Variable` in `state behavior` when activating the multi-instance body.

We could query the output variable and rely on its size. Or store the input collection size similarly as a `local variable`, we wouldn't then need to change anything to the element instance.

Reply:
>We could query the output variable and rely on its size

@mustafadagher While unexpected, and probably leading to problems, users could update the output collection variable themselves. We should not rely on process variables to determine the execution here anymore.

Reply:
@korthout please check again now

Reply:
I've discussed this solution idea with @saig0 and we concluded that:
- Regarding performance:
  - we don't actually know the performance impact the initial fix has and whether that is a problem
  - the new `MultiInstance` record theoretically also comes at the cost of some performance, e.g. serialization/deserialization of msgpack, writing and reading from RocksDB, and additional exporter load for the new record
  - there are other solutions that may be better suited to improve the performance, like caching the input collection in memory
- Regarding concurrent modification issues for the input collection:
  - there are other solutions that we anyways want to add at some point, like the introduction of immutable variables
- It's generally not a good idea to optimize prematurely

@mustafadagher With this in mind, I would actually suggest that we don't take this optimization now. Instead, let's just fix the bug with the additional read of the collection, and take our learnings by documenting them in issues:
- one to improve the performance of multi-instance
- one to avoid problems in the case of concurrent modification of the input collection

Reply:
@korthout 
For the first point, I think I presented the same argument of not knowing if saving and loading instance in the state is more performant than evaluating the expression. But your counter argument was that some expressions might need to render other expressions too to evaluate their values and that might be very slow to do multiple times.

If I understood the conclusion correctly, I will revert my last few commits saving the input collection to the event, and then I can merge this PR, is that tight?

Reply:
>If I understood the conclusion correctly, I will revert my last few commits saving the input collection to the event, and then I can merge this PR, is that right?

👍 Yes, I can provide a new review once the changes are ready